docroot/modules/custom/civictheme_migrate/src/Form/CivicThemeMigrationConfigurationForm.php

```diff
@@ -0,0 +1,500 @@
+      '#description' => $this->t('One URL each line'),
+      '#default_value' => $config->get('remote')['endpoint'] ?? NULL,
+      '#states' => [
+        'required' => [
+          ':input[name="migration_type"]' => ['value' => 'remote'],
+        ],
+      ],
+    ];
+
+    $form['remote']['actions']['#type'] = 'actions';
+    $form['remote']['actions']['retrieve_files'] = [
+      '#type' => 'submit',
+      '#value' => $this->getRetrieveFilesLabel(),
+      '#button_type' => 'primary',
+    ];
+
+    $form['configuration_files'] = [
+      '#type' => 'managed_file',
+      '#title' => $this->t('Upload Merlin extracted content JSON Files'),
+      '#default_value' => $config->get('configuration_files'),
+      '#multiple' => TRUE,
+      '#progress_indicator' => 'bar',
+      '#progress_message'   => $this->t('Uploading files...'),
+      '#upload_location' => 'private://civictheme_migrate/',
+      '#upload_validators'  => [
+        'file_validate_extensions' => ['json txt'],
+        'civictheme_migrate_validate_json' => [],
+      ],
+    ];
+
+    $form['actions']['#type'] = 'actions';
+    $form['actions']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->getSaveConfigurationLabel(),
+      '#button_type' => 'primary',
+    ];
+    $form['actions']['generate_migration'] = [
+      '#type' => 'submit',
+      '#value' => $this->getGenerateMigrationLabel(),
+      '#button_type' => 'primary',
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    if ($this->isGenerateMigrationSubmit($form_state)) {
+      $files = $form_state->getValue('configuration_files');
+      if (empty($files)) {
+        $form_state->setError($form['configuration_files'], $this->t('Configuration files are required to generate a migration'));
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    $config = $this->config('civictheme_migrate.settings');
+    if ($this->isRetrieveFilesSubmit($form_state)) {
+      $urls = explode("\r\n", $form_state->getValue('endpoint'));
+      try {
+        $files = $this->retrieveFiles($urls);
+        $existing_files = $config->get('configuration_files') ?? [];
+        $config->set('configuration_files', array_merge($existing_files, $files));
+        $this->messenger()->addStatus($this->t('Merlin extracted content JSON files have been retrieved'));
+      }
+      catch (\Exception $exception) {
+        $this->messenger()->addError($exception->getMessage());
+      }
+    }
+
+    if ($this->isSaveConfigurationSubmit($form_state) || $this->isGenerateMigrationSubmit($form_state)) {
+      $config->set('configuration_files', $form_state->getValue('configuration_files'));
+    }
+    $config->set('migration_type', $form_state->getValue('migration_type'));
```

Comment:
this code is too confusing for me. parts of it execute things, parts are saving configurations.

Reply:
Split into methods to be more explicit on what is happening.